src/ucs/sys/topo/base/topo.c

```diff
@@ -790,3 +790,49 @@ const char *ucs_topo_resolve_sysfs_path(const char *dev_path, char *path_buffer)
               sysfs_path);
     return sysfs_path;
 }
+
+static ucs_status_t
+uct_topo_iface_parse_virtual_dev(const struct dirent *entry, void *ctx)
+{
+    static const char *low_level_dev_prefix = "lower_";
+    ucs_string_buffer_t *dev_path           = ctx;
+
+    if (!strncmp(entry->d_name, low_level_dev_prefix,
+                 strlen(low_level_dev_prefix))) {
+        ucs_string_buffer_appendf(dev_path, "/%s", entry->d_name);
+        return UCS_ERR_CANCELED;
+    }
+
+    return UCS_OK;
+}
+
+ucs_status_t ucs_topo_iface_lowest_device_path(const char *dev_name,
+                                               const char *dev_type_root,
+                                               char *lowest_path_buffer,
+                                               int buffer_length)
+{
+    ucs_string_buffer_t dev_path = UCS_STRING_BUFFER_INITIALIZER;
+    const int max_depth          = 8;
+    int depth;
+    ucs_status_t status;
+    
+    ucs_string_buffer_appendf(&dev_path, "%s/%s", dev_type_root, dev_name);
+
+    for (depth = 0; depth < max_depth; ++ depth) {
+        status = ucs_sys_readdir(ucs_string_buffer_cstr(&dev_path),
+                                 uct_topo_iface_parse_virtual_dev, &dev_path);
+        if (status != UCS_ERR_CANCELED) {
```

Comment:
Is this code tested? `CANCELED` means looping again and again. Why?

Reply:
Hi @gleon99, yes, I did. It works. 

`ucs_sys_readdir` travels all the sub-directories and `uct_topo_iface_parse_virtual_dev` determines whether it's a virtual device.
1) If yes, it will return `UCS_ERR_CANCELED`, and `ucs_sys_readdir` also will break from sub-directories travelling, and return `UCS_ERR_CANCELED`. That's means a virtual device is found in the current directory and then enter the virtual device's directory for deep searching.
2) If not `UCS_ERR_CANCELED`, that means there's no virtual device in the current directory. In such case, deep searching will be ended.